core/src/runtime/execution_report.rs

```diff
@@ -0,0 +1,171 @@
+use super::Runtime;
+use std::{any::Any, collections::HashMap};
+
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+/// Errors which may occur when querying an execution report for
+/// instruction details.
+pub enum MetricRetrievalError {
+    #[error("category not found")]
+    CategoryNotFound,
+
+    #[error("opcode not found in category")]
+    OpcodeNotFound,
+
+    #[error("data type mismatch for the stored metric")]
+    DataTypeMismatch,
+}
+
+/// A report containing useful information about the execution of a program.
+/// It stores useful information about the resource footprint of a provable
+/// RISCV program, including number of cycles, instruction count, and syscall
+/// count.
+///
+/// ### Custom Metrics:
+/// An execution report contains a rich set of metrics and granular details
+/// that can be easily accessed. You can, for example, count the total
+/// number of ADD instructions that occurred during a program run:
+/// ```
+/// use sp1_core::{
+///     runtime::{Instruction, Opcode, Program, Register, Runtime},
+///     utils::SP1CoreOpts,
+/// };
+///
+/// let instructions = vec![
+///     Instruction::new(Opcode::ADD, 29, 0, 5, false, true),
+///     Instruction::new(Opcode::ADD, 30, 0, 37, false, true),
+///     Instruction::new(Opcode::ADD, 31, 30, 29, false, false),
+/// ];
+///
+/// // Setup a program and a runtime
+/// let program = Program::new(instructions, 0, 0);
+/// let mut runtime = Runtime::new(program, SP1CoreOpts::default());
+/// // Runtime must execute before reporting statistics
+/// let report = runtime.dry_run().unwrap();
+/// assert_eq!(runtime.register(Register::X31), 42);
+///
+/// // Retrieve the number of ADD operations
+/// let add_count = &report.get_metric("Arithmetic", "ADD").unwrap_or(0);
+/// assert_eq!(*add_count, 3);
+/// assert_eq!(report.cycles, 12);
+/// assert_eq!(report.total_instruction_count, 3);
+///
+/// // You can also get a count of all instruction/opcode occurrences by category
+/// let add_operations_count = report.get_total_for_category("Arithmetic").unwrap_or(0);
+/// assert_eq!(add_operations_count, 3,);
+/// ```
+#[derive(Debug)]
+pub struct ExecutionReport {
+    /// Total number of cycles occurred during program execution
+    pub cycles: u32,
+    /// Total number of instructions occurred during program execution
+    pub total_instruction_count: usize,
+    /// Total number of syscalls occurred during program execution
```

Comment:
these two statistics are not that useful, would be better to have HashMap<Opcode, usize> and HashMap<SyscallCode, usize>